includes/binary_object_upload.form.inc

```diff
@@ -93,44 +93,49 @@ function islandora_binary_object_upload_form_submit(array $form, array &$form_st
   module_load_include('inc', 'islandora', 'includes/utilities');
   $object = islandora_ingest_form_get_object($form_state);
 
-  if (!isset($object['OBJ'])) {
-    $datastream = $object->constructDatastream('OBJ', 'M');
-  }
-  else {
-    $datastream = $object['OBJ'];
-  }
-  $file = file_load($form_state['values']['file']);
-  $datastream->setContentFromFile($file->uri, FALSE);
-  if ($datastream->label != $file->filename) {
-    $datastream->label = $file->filename;
-  }
-  if ($datastream->mimetype != $file->filemime) {
-    $datastream->mimetype = $file->filemime;
-  }
+  if ($form_state['values']['file']) {
+    $file = file_load($form_state['values']['file']);
+    $datastream = isset($object['OBJ']) ?
+      $object['OBJ'] :
+      $datastream = $object->constructDatastream('OBJ', 'M');
+
+    $datastream->setContentFromFile($file->uri, FALSE);
 
-  if (!isset($object['OBJ'])) {
-    $object->ingestDatastream($datastream);
+    if ($datastream->label != $file->filename) {
+      $datastream->label = $file->filename;
+    }
+
+    if ($datastream->mimetype != $file->filemime) {
+      $datastream->mimetype = $file->filemime;
+    }
+
+    if (!isset($object['OBJ'])) {
+      $object->ingestDatastream($datastream);
+    }
   }
 
   if ($form_state['values']['supply_thumbnail']) {
     $thumbnail_file = file_load($form_state['values']['thumbnail_file']);
+
     if ($form_state['values']['scale_thumbnail']) {
       islandora_scale_thumbnail($thumbnail_file, 200, 200);
     }
+  }
```

Comment:
Think this is meant to go at the end of all the `$tn` logic? In its current state, we're constructing and ingesting a TN datastream where one may not exist, and playing around with a `$thumbnail_file` that also may not exist.